src/objective-c/RxLibrary/GRXForwardingWriter.m

```diff
@@ -80,22 +76,38 @@ - (GRXWriterState)state {
 }
 
 - (void)setState:(GRXWriterState)state {
+  GRXWriter *copiedWriter = nil;
   if (state == GRXWriterStateFinished) {
-    _writeable = nil;
-    [self finishInput];
+    @synchronized(self) {
+      _writeable = nil;
+      copiedWriter = _writer;
+      _writer = nil;
+    }
+    copiedWriter.state = GRXWriterStateFinished;
```

Comment:
Why not change state inside `@synchronized(self) {}` without copying the writer?

Reply:
Deadlock prevention. Locking order is always writer->writeable. Cannot call writer's function inside a @synchronized block.

Reply:
Oh, you are changing it to atomic. I see.